django/contrib/admin/views/autocomplete.py

```diff
@@ -15,6 +17,34 @@ def get(self, request, *args, **kwargs):
             pagination: {more: true}
         }
         """
+        self.term = request.GET.get('term', '')
+
+        try:
+            app_label = request.GET['app_label']
+            model_name = request.GET['model_name']
+            field_name = request.GET['field_name']
+        except KeyError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
+        try:
+            source_model = apps.get_model(app_label, model_name)
+        except LookupError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
+        try:
+            self.source_field = source_model._meta.get_field(field_name)
+        except FieldDoesNotExist:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
+        try:
+            remote_model = self.source_field.remote_field.model
+        except AttributeError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+        try:
+            self.model_admin = self.admin_site._registry[remote_model]
+        except KeyError:
+            return JsonResponse({'error': '403 Forbidden'}, status=403)
+
```

Comment:
Hey @codingjoe — so I think I would restructure this section slightly, through to line 59. 

* I'd create a method to configure three attributes on self: `source_field`, `model_admin`, and `to_field_name`. It'd need to take the `request`. In there I'd include all the `try:except:` blocks and raise an `AutoCompleteSetupError` if any of them occur. 
* Outside that method I would have a single `try:except:` to log the error, and return the JSON 403 response. (I'm thinking about how I'd debug this if there's no logging, and all I have to go on is a 403...) 

I think we should also move the `has_perm() -> 403` check **above** the `get_search_fields() -> 404`. (That last if nicely informative for users but we shouldn't hand out that info to someone without the permissions to access the modeladmin at all… 🤔)

Reply:
Very good idea, I moved the code into a separate method. I further agree with the problem of debugging potential errors. When I originally submitted the patch we switch to a 403 JSON response, to have a consistent response content type. However, that is not necessary, since Select2 does not care about the content type. If the response code is 4xx or 5xx it will always display the same message and not even try to decode the content. Therefore, I believe it's best to simply raise a `PermissionDenied` and let the middleware handle logging. I added proper `raise from` clauses, to provide users with additional information.

I kept the `has_perm` call in the `get` method, to make sure it doesn't get lost and I don't find it logically connected to the rest.